### PR TITLE
[ShallowRenderer] Queue/rerender on dispatched action after render component with hooks

### PR DESCRIPTION
**Pull request by rodrigopr**
This should help make shallow render a viable option for testing components that use hooks.

### CODE REVIEW

**Comment by rodrigopr** [comment]
@gaearon, @trueadm, @acdlite, could you review this? (sorry to ping like this)
Being able to trigger hooks state changes (useState, useReducer) after render will allow us to keep most of our shallow tests intact if we migrate components to use hooks.

File: packages/react-test-renderer/src/ReactShallowRenderer.js [hunk 1]
```diff
@@ -384,7 +384,7 @@ class ReactShallowRenderer {
         'an infinite loop.',
     );
 
-    if (componentIdentity === this._currentlyRenderingComponent) {
+    if (componentIdentity === this._previousComponentIdentity) {
```
Comment (gaearon) [inline packages/react-test-renderer/src/ReactShallowRenderer.js]: Why is this change correct? It disagrees with what the comment inside says ("This is a render phase update"). What was it supposed to mean? Why was it changed?

**Comment by gaearon** [comment]
I reverted in https://github.com/facebook/react/pull/14839 — I appreciate the PR but I wanted to cut a patch soon and it's not obvious to me that this change is correct. (Although you're right the existing code is not either.)

This code was originally written with no outside state updates in mind. It was written as 1:1 to the server renderer. It might need a larger refactoring to work with updates at arbitrary points in time.

At the least, we should rename the variables and rewrite the comments that indicate these are only render-phase updates. With this change the logic doesn't match the comments and variables, and I'm worried this will make the code so confusing that nobody will be able to fix further problems in it.

**Comment by gaearon** [comment]
(I'd be happy if you could take this work. Let's just ensure the code matches the intention, and that we examine it all instead of minimal tweaks to "make it work".)

**Comment by gaearon** [comment]
I filed https://github.com/facebook/react/issues/14840 to track this.

File: packages/react-test-renderer/src/ReactShallowRenderer.js (hunk 1 above)
Comment (rodrigopr) [reply packages/react-test-renderer/src/ReactShallowRenderer.js]: Yeah, I missed updating the comment =/
Before this change `_dispatchAction` ignored calls after the render returned, this checks if its the same component and queue the action.

**Comment by rodrigopr** [comment]
No problem @gaearon, I'll move the discussion to the issue.

**Comment by gaearon** [comment]
Thanks!

**Comment by trueadm** [comment]
Thanks for digging into this deeper. I have a larger fix ready that should handle this case properly.

**Pull request merged**